tests/unit/utils/test_io_utils.py

```diff
@@ -0,0 +1,128 @@
+
+
+def test_read_file_contents_as_string_works(tmp_path):
+    """Tests reading from file."""
+    file_path = os.path.join(tmp_path, "test.txt")
+    io_utils.write_file_contents_as_string(file_path, "aria_best_cat")
+    assert io_utils.read_file_contents_as_string(file_path) == "aria_best_cat"
+
+
+def test_read_file_contents_as_string_raises_error_when_file_not_exists(
+    tmp_path,
+):
+    """Tests reading from file."""
+    file_path = os.path.join(tmp_path, "test.txt")
+    with pytest.raises(FileNotFoundError):
+        io_utils.read_file_contents_as_string(file_path)
+
+
+def test_copy_dir_works(tmp_path):
+    """Tests copying directory."""
+    dir_path = os.path.join(tmp_path, "test")
+    fileio.create_dir_if_not_exists(dir_path)
+    file_path = os.path.join(dir_path, "test.txt")
+    fileio.create_file_if_not_exists(file_path, "some_content_about_aria")
+
+    new_dir_path = os.path.join(tmp_path, "test2")
+    io_utils.copy_dir(dir_path, new_dir_path)
+    assert os.path.exists(new_dir_path)
+    assert os.path.isdir(new_dir_path)
+    assert os.path.exists(os.path.join(new_dir_path, "test.txt"))
+    assert os.path.isfile(os.path.join(new_dir_path, "test.txt"))
+    with open(os.path.join(new_dir_path, "test.txt"), "r") as f:
+        assert f.read() == "some_content_about_aria"
+
+
+def test_copy_dir_overwriting_works(tmp_path):
+    """Tests copying directory overwriting."""
+    dir_path = os.path.join(tmp_path, "test")
+    fileio.create_dir_if_not_exists(dir_path)
+    file_path = os.path.join(dir_path, "test.txt")
+    fileio.create_file_if_not_exists(file_path, "some_content_about_aria")
+
+    new_dir_path = os.path.join(tmp_path, "test2")
+    fileio.create_dir_if_not_exists(new_dir_path)
+    file_path = os.path.join(new_dir_path, "test.txt")
+    fileio.create_file_if_not_exists(file_path, "some_content_about_blupus")
+
+    io_utils.copy_dir(dir_path, new_dir_path, overwrite=True)
+    assert os.path.exists(new_dir_path)
+    assert os.path.isdir(new_dir_path)
+    assert os.path.exists(os.path.join(new_dir_path, "test.txt"))
+    assert os.path.isfile(os.path.join(new_dir_path, "test.txt"))
+    with open(os.path.join(new_dir_path, "test.txt"), "r") as f:
+        assert f.read() == "some_content_about_aria"
+
+
+def test_copy_dir_throws_error_if_overwriting(tmp_path):
+    """Tests copying directory throwing error if overwriting."""
+    dir_path = os.path.join(tmp_path, "test")
+    fileio.create_dir_if_not_exists(dir_path)
+    file_path = os.path.join(dir_path, "test.txt")
+    fileio.create_file_if_not_exists(file_path, "some_content_about_aria")
+
+    new_dir_path = os.path.join(tmp_path, "test2")
+    fileio.create_dir_if_not_exists(new_dir_path)
+    file_path = os.path.join(new_dir_path, "test.txt")
+    fileio.create_file_if_not_exists(file_path, "some_content_about_blupus")
+
+    with pytest.raises(FileExistsError):
+        io_utils.copy_dir(dir_path, new_dir_path, overwrite=False)
```

Comment:
could perhaps add another assert at the end to make sure that the content is still "some_content_about_aria"

Reply:
ah sorry I mean to add an assert that ensures that the old file was not overwritten, so in this case we would need to assert the contents of `file_path2`